### DISCUSSION TITLE
Allow GIFs in Bluesky

### DISCUSSION THREAD

**Discussion by dholms** [bluesky-social/atproto#1157] [Bluesky Lexicons]
They're fun & right now we only support jpeg & png

**Comment by bnewbold** [comment]
I think the common consensus these days is to convert to video though. GIF animations are often huge and would get mangled if resized on upload to fit in, eg, 1MB limit. It is basically video, in terms of bandwidth, UX considerations, etc.

While we are at it, could also consider webp and APNG animations. I don't have any market share numbers about how prevalent those actually are, so maybe not a priority.

But, totally agree fun!

**Reply by olivercarl** [reply]
I think the average user doesn't care what in format the "animated pictures" are delivered. Convert it to whatever is best for storage and delivery.

**Reply by KarlLivesey** [reply]
the easiest is just just convert the gif to a video for the user they don't care on the format usually

**Comment by TodePond** [comment]
+1 to this. They're needed by the '#BuildInPublic' users. ie: Developers/artists/creators post short work-in-progress gifs of what they're working on.

I think it would be worth prioritising this, because without it - those users can't migrate their posting over.
eg: https://twitter.com/steveruizok
eg: https://twitter.com/lorenschmidt

**Comment by nazgum** [comment]
im an indie game developer, twitter has (or at least had) a really popular community of indie game devs who share gifs of their progress on their games.

would be nice if bluesky/at supported them so this type of content can be posted there too

**Comment by Scrxtchy** [comment]
> While we are at it, could also consider webp and APNG animations. I don't have any market share numbers about how prevalent those actually are, so maybe not a priority.

webp has rather great usage support across browers given that it is also supported on ios versions 14 and above and android 4 and above. apng has greater support for older devices/browsers, but it's compression methods aren't as favourable to webp.
Another project I have ties to considered apng for animated user avatars, but it didn't have much favours over the output file sizes that were given when compared to webp

The avif image format has a growing but not comparably high usage/compatibility in browsers, it's based on the AV1 video format and compresses better than webp in some cases, but getting data on how well these are support across non-browser experiences didn't bear any fruit

I think other data like it's encoding and decoding performance would be important for both client and server, but I don't have good data on these

**Comment by Nezteb** [comment]
![so-sayweall-battlestart-galactica](https://github.com/bluesky-social/atproto/assets/3588798/3b0e4973-a219-4b55-a1f0-77decedd714c)

**Comment by Zero3K** [comment]
+1 to this suggestion.

**Comment by Zero3K** [comment]
Any update on this?

**Reply by dholms** [reply]
not on the near-term roadmap - we simply have higher priorities

**Reply by TodePond** [reply]
Can't post to bluesky until it does :'(

**Reply by Zero3K** [reply]
So, it will be late next year. Someone just needs to make a pull request so the team can focus on the moderation code.

**Reply by yetisyny** [reply]
It would be simplest just to implement animated GIF support directly without converting it to video like Twitter has always done and which Mastodon also does, but this could lead to bigger files and slow things down (more bandwith and client CPU usage).

@bnewbold proposed converting to video. This is what top animated GIF sites like Tenor and Giphy do, converting to MP4 files with H.264 video and no audio track.

If we converted to multiple formats like also coverting to newer codecs like VP9, AV1, and HEVC, that would require extra computational power and disk space on the server and sniffing what formats a browser or device supports and just be excessively complicated.

So for animated GIFs, if we convert to video, best to just stick with MP4 containers with the H.264 codec.

Just having the original animated GIF unchanged would be simpler but admittedly this uses more bandwidth when people browse the site/app. Also displaying an animated GIF is computationally intensive on mobile devices that run on battery and wastes CPU and battery, while H.264 video is hardware accelerated on all Android and iOS devices and universally supported by all modern web browsers.

If we do video conversion, it’s easiest to implement by calling FFmpeg to convert the animated GIF file to MP4 using the x264 codec. Because of licensing issues, you can’t directly integrate FFmpeg into anything not under GNU GPL so that means it can’t be a compile-time dependancy but can be a run-time dependency. So an external binary would be called to do the file conversion to MP4.

Sites like Tenor and Giphy display in MP4 by default but give you the option to view and download the original animated GIF file. Having an interface to view and download the original animated GIF file, if we show it as a video instead, would be very useful to users. I think it best to keep the original animated GIF file for a post on the server and not delete it after the conversion to MP4. It is likely people would want to be able to see and download the original file to avoid losing quality.

Adding support for other input and output formats could be added later, but it’s best to keep things simple at first, I think.

MP4 with H.264 works well across all browsers and devices and has been standard for years, and if we tried using a newer codec, we’d still need it as a fallback format on devices and browsers that don’t support the newer one.

Of course the simplest solution would be just upload animated GIF files unchanged and display them unchanged. I think Twitter and Mastodon do that. It’s not ideal on mobile phones with limited data and battery, but it’s better than not supporting animated GIFs at all. If we just go with animated GIFs being unchanged, we could add an option on the mobile app and the website settings to enable or disable animated GIFs, where disabling would just show you a smaller GIF or PNG file of the first frame.

I’m fine with either way and would just like to see animated GIF support implemented soon. Thanks.

**Reply by bookshop** [reply]
Hi! I'm just a lowly bluesky user here trying to figure out what the ETA is on this is, so i'm really sorry if this comment is out of place or seems intrusive. i just really want to give the perspective of a longtime member of fandom on twitter. (since roughly 2009.)

the international fan community on twitter desperately needs a stable twitter alternative to migrate to, one that's easy to set up and navigate. i've been trying very hard to recruit fans to bluesky, but it's extremely difficult — even embarrassing — to ask major fandom influencer accounts to migrate to bluesky when it doesn't have .gif (converted format is fine) and video upload capabilities (.mp4 prolly ideal for most users), especially when there's no ETA for when they'll be added. It is really impossible to overstate how integral gifs and fanvids are to the culture of fandom on twitter, and by not having this feature you're effectively losing the potential to easily recruit millions of fans from around the world to the platform. 

i know that's probably not a priority to the bluesky devs, but there are countless fan communities and cultures residing on twitter that will likely be lost if twitter collapses and nothing exists to replace them, because the nature of an international community like this is that without quick simple access to an easy alternative — and crucially, an alternative they've been encouraged to join from other fans — members of fandom will just delete their content and cease to be active, losing all their content and all their connections and community.

i know that sounds like a lot of pressure to put on .gif and video capability, but it's entirely accurate. so just want to say: please think of us, the fans! we need a (gif-friendly) lifeline!

**Reply by Scrxtchy** [reply]
I don't think there's an ETA on this as there's other more features that the team are working on, the biggest being the federation of atproto PDS which is being actively worked on in a sandbox environment.  Even with this in mind, getting animated content support into the network is more than just allowing users to upload the content, having the types written and displaying it on the website, but it also must integrate appropriately with the content labelling and moderation filters though Hive or otherwise

**Reply by yetisyny** [reply]
Yeah agreed 100% @bookshop, couldn’t have said it better myself. I hope animated GIF support is added ASAP.

Someone needs to code it and submit a pull request that adds this feature. You don’t have to be a Bluesky developer to do it, anyone with a GitHub account can add animated GIF support to Bluesky if they have the right skills and enough time to spend on it to get it working.

**Reply by amoNoot** [reply]
I would _seriously_ begin considering giving higher prioritization to the reason people literally browse and join social media platforms for. The very idea that it is not even on the near-term roadmap is honestly troubling, from any perspective.

**Comment by FaeAlicia** [comment]
Huge +1 for this feature.

> It would be simplest just to implement animated GIF support directly without converting it to video like Twitter has always done and which Mastodon also does, but this could lead to bigger files and slow things down (more bandwith and client CPU usage).
> 
> @bnewbold proposed converting to video. This is what top animated GIF sites like Tenor and Giphy do, converting to MP4 files with H.264 video and no audio track.

My understanding is that Twitter has done the conversion to video for quite a long time. Back when I used it over 5 years ago, I once uploaded a video that lacked audio and it was treated as a gif, getting both the looping playback treatment and that little "GIF" text that shows in the bottom-left corner.

Also, any time I've tried to download a gif from Twitter, it's given me an MP4 video that uses H.264 and lacks an audio track.

I also took a random gif on my phone and with Termux, used ffmpeg to convert it to H.264. Took just a couple seconds, even on a low-power phone and the file size went from >6MB to <250KB.

So I'd say yes, just do the conversion to H.264 and save everyone the resources in the end. If a phone can do it in a couple seconds, no one's gonna notice it being converted live by a server and then served back to them in the composition's image preview area.

**Reply by Chad90b** [reply]
as long as the forced/automated conversion is to a lossless format so the art posted isnt irreversibly damaged i can imagine all users would probably be okay with it. as soon as we start introducing lossy conversion thats where real pushback could occur. im sure passionate art fans are a minority, but i just cant imagine why anyone would want to spend time creating a planform that irreversibly alters the content posted. hopefully this is something thats taken into consideration with this big debate of simply...adding gif support.

**Comment by dyoung404** [comment]
Also a +1 for this with the MP4 container and H.264 codec (for now). If the goal is to have folks migrate from Twitter and other services like Mastodon and to grow bsky, uploading gifs and short videos should definitely be part of the very-near-term roadmap. I fully understand the complexity and resourcing involved, but this is also pretty much a core feature of every single social network / app that bsky is competing with.

Love what you folks are doing!

**Comment by Supernorn** [comment]
+1 to GIF support - I'm a game developer who primarily uses twitter to share progress on our titles. GIFs have been essential for this.

GIF support should be an urgent priority if users are to consider Bluesky as a serious replacement to Twitter going forward. 
Right now, Bluesky is simply unable to compete with other social media platforms with this core feature missing.

**Comment by yongminparks** [comment]
just passing by and dropping +1

**Comment by gingerbeardman** [comment]
I'm also a game developer. 

I've been able to select GIFs from my Photos when running the iOS app on macOS. But when it is ~~uploaded~~ downloaded it's a static file of the first frame. Maybe a preview image? 

So other than the technical aspect of which file format, there's also the user interface and user experience considerations. On Twitter and Mastodon I can add images by pasting a link or image data from the clipboard, dragging and dropping, or the system picker.

**Comment by pem884** [comment]
I'm a new-ish user who has wished for GIF support on several occasions already; dropping by in support of the functionality.

**Comment by todordvv** [comment]
+1

need GIFs asap!

**Comment by gingerbeardman** [comment]
> I've been able to select GIFs from my Photos when running the iOS app on macOS. But when it is uploaded it's a static file of the first frame. Maybe a preview image is being uploaded?

I double checked this: it is in fact an upscaled PNG that gets ~~uploaded~~ downloaded.

**Reply by bnewbold** [reply]
just to clarify this small thing: I believe what is probably happening is that the original GIF is uploading as-is (byte-for-byte), and then when a thumbnail is generated it is a non-animated PNG.

**Reply by gingerbeardman** [reply]
Oh, thanks for that insight!

**Reply by TheTechmage** [reply]
You gave me the idea to upload an animated png and the thumbnail downloaded as a jpeg. I was really hoping that animated pngs might just be a "secret way to get 'gif' support in the interim" xD

**Comment by sudoWright** [comment]
+1 GIF support! Thank you.

**Comment by ICF19XX** [comment]
+1 GIF support! Thank you!

**Comment by TodePond** [comment]
+1 GIF support! Thank you!

**Comment by penx** [comment]
![+1 GIF support! Thank you!](https://github.com/bluesky-social/atproto/assets/1027024/8ba3ce40-52f9-4271-99d8-df7d5902a4e3)

**Comment by DavidBuchanan314** [comment]
I'm not the github discussion police, but I do get notifications for activity here, so I'd appreciate fewer redundant +1 comments. The upvote button exists!

<img src="https://media.tenor.com/kgc_-cfOicoAAAAC/angry-raging.gif"  width="40px" />

**Reply by TodePond** [reply]
I apologise about that. Would you like gif support by the way?

**Comment by warcode** [comment]
You've got more artist refugees incoming due to AI theft elsewhere so the time for this is now.

**Comment by TodePond** [comment]
It's clear there's a lot of support for this. Maybe someone could make an issue?

**Reply by Scrxtchy** [reply]
This used to be the issue but it was migrated into a discussion page #519

**Reply by Zero3K** [reply]
Which is this discussion page.

**Reply by TodePond** [reply]
Ok thanks for the info, I'll make an issue then, unless someone beats me to it

**Comment by olivercarl** [comment]
Besides the general support for GIFs/APNG/... (not sure if this is being discussed elsewhere):
This content is normally not uploaded by the user. Rather, the user chooses it from a library (searchable by keywords). I think without this aspect, most users would be disappointed.

**Reply by bookshop** [reply]
i do think a GIF library would be great, but also plenty of people do upload their own gifs — it really just depends on the context. it's not that hard to search the web, grab/save (or copy and then paste) a gif and upload it that way.  just having the functionality on site is a huge start.

**Reply by TodePond** [reply]
Can also confirm I post gifs manually on mastodon and X. It's a common practice in the "build-in-public" community

**Comment by juniberry** [comment]
+1 for GIF support!!! Please!!

**Comment by TodePond** [comment]
It seems like there is still an open issue about GIFs here: https://github.com/bluesky-social/social-app/issues/1047

Be sure to be a part of that conversation!

**Comment by middleone20** [comment]
+1 GIF/Video Support. Bluesky is gaining traction and it would be a shame for new users to revert back to other alternatives due to such a crucial feature missing. I am especially thinking of artists, who often upload process videos or animations they've created to social media. Thanks for all the great work on Bluesky!

**Comment by Arsenic110** [comment]
Considering how long this has been open - has anything been done, is anything planned, or has anything been announced?
This is quite frankly a very basic feature I was surprised to find out isn't (hopefully yet?) implemented.

**Comment by TodePond** [comment]
@Arsenic110 unfortunately not, although some third-party apps have been picking up the pieces, like Graysky and deck.blue

They have done quite a good job despite the limitations they're working with (eg: file size limit). At the very least, I think those limitations could be improved.

+1 GIF support! Thank you!

**Reply by inkygirl** [reply]
When/if Bluesky adds gifs, I am really hoping they enable people to access gifs from GIPHY. Graysky and deck.blue only access Tenor. I'm a GIPHY artist and don't find Tenor nearly as user-friendly for a creator. Fingers crossed!

**Reply by surfdude29** [reply]
You're in luck, please see https://github.com/bluesky-social/social-app/issues/2314 and https://github.com/bluesky-social/social-app/pull/2353

At this stage it's embed support rather than full built-in support, but it's still an important step forward

**Reply by TodePond** [reply]
oh no... uploadable gifs are the requested thing here. not embedded ones

**Reply by haileyok** [reply]
So take what I say with a grain of salt, as obviously design decisions are not up to me. And I certainly think that actually uploading user GIFs is definitely the ultimate solution long term.

However, having the embed support is probably the *most* important aspect of having GIFs available. The vast majority of times, the GIFs that people use are going to be already created ones that are hosted elsewhere. GIFs on Twitter for example are all offered through GIPHY, Discord uses Tenor, etc.

I think it's important to note too that even though this is labeled as an "embed", its actually playing the actual GIF file, not embedding through Giphy/Tenor. There is virtually no difference here between user-uploaded GIFs and the Giphy "embeds" aside from the fact that you are having to link the GIF instead of uploading it.

Again, don't make design decisions, but with this support it is very viable to implement a Giphy search that operates in the same way that Discord, Twitter, etc. all do and without the immediate need to support and handle the uploading of many large files.

Just my two cents.

**Reply by haileyok** [reply]
FWIW, I do understand that there's a big need for GIFs that are not "meme" related. I.e. I think you've mentioned wanting to show some of your work perhaps through GIFs, and I think that's a super important thing to have available.

**Comment by Zero3K** [comment]
Another +1.

**Comment by unluckychloe** [comment]
adding my +1 to this, im a game developer and the lack of gifs here means i literally cannot move over from twitter. nearly every single offline convo i've had about this site as a twitter alternative ends with "too bad there's no gifs/video". seems like the team isn't interested in prioritizing it but seriously, you have no idea how much this is holding the site back from being a viable alternative

**Comment by TodePond** [comment]
That seems like an important feature! Perhaps there should be an issue for it

**Reply by surfdude29** [reply]
There is for the app itself here:

https://github.com/bluesky-social/social-app/issues/1047

**Reply by Scrxtchy** [reply]
lol

**Comment by inkygirl** [comment]
HURRAAAAYYYYYYYYYYY!!!!!

Thanks for the heads-up!

Debbie

Debbie Ridpath Ohi - Children’s book author & illustrator
Website ~ Twitter ~  Instagram (art) ~ Bookstagram - Substack - BlueSky

**Comment by TodePond** [comment]
<img width="1064" alt="shapes at 23-12-28 23 35 06" src="https://github.com/bluesky-social/atproto/assets/15892272/2333d753-55f7-47ab-9258-cb7de0027ebc">

**Reply by ArchivingToolsForWBM** [reply]
And the gifs gets converted into a jpeg that shows only the first frame of the image

**Comment by est** [comment]
should I mention I have a gif avatar on TWTR ?  (lookup my handle same as my github)

**Reply by Zero3K** [reply]
What does that have to do with this discussion?

**Reply by est** [reply]
by allowing gif/vid in posts, naturally the next step is gif in avatars.

**Reply by Zero3K** [reply]
Oh, then that makes sense.

**Comment by Jon-dog** [comment]
Another +1 from me, just uploaded a gif and was shocked to find it displayed only a single frame. Not even attempting to convert it to some modern format or even just telling me gifs aren't supported feels misleading

**Comment by mozzius** [comment]
We've shipped Tenor integration now!

**Reply by michelmohr** [reply]
That's not... that's not what we meant.

**Reply by Scrxtchy** [reply]
it's still on the roadmap, please be patient

**Reply by tinspin** [reply]
Your roadmap has the wrong priorities.

Uploaded animated gifs is in position 1 before you even start a social media project.

Position 2 is make a system that allows animated gifs to have audio attached along side it.

Position 3 is convert all other media to this superior format with sweet dithering.

Position 4 is remove all other video formats. XD

**Comment by shmerl** [comment]
Also better than gif and apng - allow support for avif images which also have animation.

**Comment by codeMotif** [comment]
I'm a pixel-art animator who wants to participate in Bluesky. My biggest and most heavy gifs are less than 500 KB, and most of my gifs are 100 KB on average. And I'm being prohibited from posting them. Gifs are a native feature of every single browser. Rendering them is extremely easy. Embed the image, and you're done. Set a reasonable upload limit and let the people do what they will with it. Maybe some of them will fumble and the gif will look funny. That's how it goes. If you're serving static images, you can serve a gif by putting the gif in the markdown my browser gets served. A clickthrough or a setting for epileptic accessibility, at the absolute maximum, even though gif-handling accessibility tools exist for many browsers to address that need anyway.

Enable the gif filetype. Do not compress it, resize it, or otherwise alter the file uploaded by the user in any way besides renaming it. Then serve it to the browser and let the browser render it. It's incredibly easy and demands nothing besides not overhandling it. Anything less than this course of action is aesthetic garbage that prioritizes visual branding over user experience or user growth, and I won't use this site at all until it works that way.

**Comment by tutipila** [comment]
Hello! I just uploaded a GIF to Tenor and its already appearing on other apps, but still isn't appearing on Bluesky gif search. Is there any cache to refresh?

**Reply by michelmohr** [reply]
Please don't bring tenor into this thread, it's hard enough to convey that we want to upload custom gifs directly to the platform.

**Comment by ninthpower** [comment]
As a developer and designer myself, I think you'll find users will jump through any hoops to post animated content if quality is preserved. Users will jump through whatever conversion/codec hoops they need to.

If users have to host their content, they'll find a way.

If users have to pre-convert their content, they'll find a way.

**Comment by KennedyRichard** [comment]
I'm glad Bluesky is about to get video support, but I'd like to give a +1 to this.

I'm a open-source maintainer and GIFs (or MP4s that behave like GIFs) are very important to my work. Although videos are crucial as well, being able to show a small piece of animated content about an app or game is very important. Specially when that piece plays automatically or at least when the mouse hovers like it is done on mastodon/fosstodon.

It helps a lot with showcasing features/animations. Even more so than videos on many occasions cause the instantaneous nature of GIFs (or MP4s that behave just like GIFs) means that that short animation is presented right away to the audience and in a short interval of time, which is suitable considering the short attention span of us internet users.

If you are hesitant about supporting user-uploaded GIFs on the platform, at least allow short MP4s with autoplay and no audio. They'd serve the same purpose, use much less space and I'm sure creators like myself would not mind using the MP4 format. It is probably the format in which they captured the media originally anyway (that or MKVs).

**Reply by haileyok** [reply]
We have a path forward on doing this, it just isn't making it into this very first iteration on video. I suspect it will be a fairly fast follow though.

**Reply by KennedyRichard** [reply]
Thank you for the prompt reply and clarification.

**Reply by codeMotif** [reply]
> We have a path forward on doing this, it just isn't making it into this very first iteration on video. I suspect it will be a fairly fast follow though.

It's going to be integrated with video? Will this involve obligated compression or conversion? The current compression-decay and filesize-increase issues for small PNGs has me worried.

**Comment by Skjerlocked** [comment]
+1  for gif support.
Thanks

**Comment by Momdere** [comment]
I'm a little late to the party on this discussion....BUT....there was just a massive amount of new users who signed up on BlueSky, over half a million people, myself included, and I keep seeing the same topic pretty frequently pop up about not being able to upload GIFs yet, which for many it seems to be a deal breaker or at least a turn off even though they seem to WANT to make the swap over to bsky for good, just not YET without certain features.

I feel like GIF upload support in some form could be beneficial, especially now, even if it doesn't feel incredibly "important" or it feels small in some way, it definitely doesn't go unnoticed by the users. I feel like it's on everyone's radar who is helping with development, so I can only hope we do see some official progress on much loved and wanted features like allowing uploads of animated GIFs soon :)

**Reply by surfdude29** [reply]
I know it's far from ideal, and of course actual support for uploading GIFs is necessary and will be much-appreciated by many when it arrives... but I feel like there's a pretty easy workaround for now that will give a very similar experience: Simply convert your GIF to a video with no audio track and upload that.

But maybe I'm missing something obvious as to why that isn't a good interim solution? 🤔

**Reply by Momdere** [reply]
I do think that's one solution for now for those who are willing to take the time to convert to video. It's quick and simple, but in the long run it would be nice to have the ease that other platforms have with uploading. I feel like the parts are there, you can even upload animated GIF data in blob format TO bsky servers and have it playback via a third-party client app, so I do wonder if maybe it's not too far off from being a reality, but there could be reasons it hasn't just been allowed YET.

Basically I think people like ease of use, if they have to go somewhere else off-site to do something with "extra steps" more or less causes em to wanna post "aww, no GIF support yet?" rather than just convert it.

Just I thought it might be good to at least have a note of the giant wave of people coming in. Something like this would definitely help keep more of em interested in staying full time

**Reply by bookshop** [reply]
@surfdude29's comment is an important one because it fundamentally misunderstands what we're trying to communicate about GIFs in this thread. "Simply convert your GIF to a video with no audio track and upload that" would be fine if we were talking about a single person with a single gif they want to share.

We aren't talking about a one-off gif. We're talking about entire subcultures who _thrive_ on gifs as a routinized part of the language of that subculture. We're talking about millions of people in fandom and standom on twitter, in queer communities and kink communities and nerd communities, who have all formed and evolved around being able to communicate using gifs to share information specific to that culture. We're talking about an entire "gif culture" that cannot relocate to Bluesky because there is no infrastructural support for it. And these are gifs that for the most part *do not exist* on Tenor because they are again specific to the language of individual communities and subcultures and no one outside that community cares about them; these aren't general reaction gifs, these are *modes of communication and information sharing.*

It's really frustrating, I don't know how I can make it clearer what a loss this is. 

My fandom is part of an enormous international fan community on Twitter, by the millions and millions of users, and every single time someone comes over from Twitter to check out how things are evolving here, the very first thing they say is "oh, we still can't upload gifs" and then they go away for another 3 or 6 months. 

This is not a nice-to-have, it is THE thing that people are waiting on to come to Bluesky.

**Comment by suwacrab** [comment]
+1, though as mentioned above in this thread by the fellow pixel artist, I'm very much on the side that the original .gif file should be kept preserved, ***lossless***, for viewing. Conversion to h.264 can be very destructive, often requiring the artist to resize their artwork to obscene sizes to lessen the amount of artifacts. This (along with automatic jpeg conversion, though that's another topic) is one of the reasons i see many other animators, pixel art or not, decide to still remain on twitter, even with the much better community that could be here.

**Comment by gabrieloc** [comment]
I've left a similar comment in another GIF-support issue, but this is a table-stakes feature and a primary medium consumed on the other platform by many of its users. Lack of GIF support makes it unviable for many. Please add support for uploading GIFs 🥺

**Comment by vrad-exe** [comment]
Wait so are gifs not supported at all? They're accepted as a file type and animate in the post dialog but then it just didn't when making the actual post which is really confusing

**Reply by spider623** [reply]
they are handled as videos, like reddit, twitter, etc. do, videos are smaller in size than gif, the alternative is to enable av1 and avif support on the site and be done with it since animated avif are small as hell, we are talking less than 10MB for a minute long loop and av1 videos, can run with no artifacts on crazy low bitrate

**Reply by vrad-exe** [reply]
That's not what was happening for me though, it just posted a static image, despite the fact that it appeared animated in the dialog where you write the post

**Reply by spider623** [reply]
odd, gif work for me on ios... did test avif but it converted it on single frame PNG

**Comment by tinspin** [comment]
For anyone look to a practical solution to this problem you can use threads to upload your gif and then take the mp4 they generate because it has less compression than twitter or any other system and in turn upload that to bluesky. Seems to work = bluesky does not compress it to smithereens.

**Comment by jlandahl** [comment]
As someone who just started using Bluesky seriously in the last few days (along with millions of others), I was really surprised that I couldn't upload a GIF as I've done for years on other platforms. Took me a few posts before I realized it just isn't supported, and then found this nearly two-year-old issue showing how long this has been requested.

A quick web search turned up [a simple and effective ffmpeg command](https://unix.stackexchange.com/a/294892/34790) that does a great job of converting GIFs to MP4s while also optimizing load time and file size:

```sh
ffmpeg -i animated.gif -movflags faststart -pix_fmt yuv420p -vf "scale=trunc(iw/2)*2:trunc(ih/2)*2" video.mp4
```

Bluesky folks, I know there's probably a very long backlog, but this or its equivalent with an FFMPEG library would be incredibly simple to implement. Not having GIF upload support is a really glaring omission in this day and age, especially when the solution is so simple.

For anyone interested in the meantime, you can mass-convert a directory of GIFs with the following command on Linux:

```sh
for f in *.gif; do ffmpeg -i "$f" -movflags faststart -pix_fmt yuv420p -vf "scale=trunc(iw/2)*2:trunc(ih/2)*2" "${f%.*}.mp4"; done
```

Edit: Please note that I don't know anything about image or video formats and just wanted a quick solution. Converting to MP4 worked for me and had the bonus of drastically reduced file sizes. Bluesky auto-plays and loops MP4s, so it works mostly as expected except that there's an annoying pause at the end of the first loop. If there's a better target format than MP4 I'd be happy to hear it.

**Comment by shmerl** [comment]
Using videos instead of gifs in general is always a good idea, but gifs in the browser behave differently, simply always looping by default. I suppose that's the main appeal. Other than that, there is zero point to ever use gifs.

**Reply by jlandahl** [reply]
I noticed that Bluesky loops MP4s, so converting from GIF to MP4 works reasonably well. There's an annoying pause at the end of the first loop, though, but after that it's pretty smooth.

**Reply by shmerl** [reply]
Does bluesky even allow uploading videos? I haven't tried recently, but it was impossible a while ago.

**Reply by jlandahl** [reply]
Yes, there are three relevant buttons when creating a post: one for images, one for video, and one for GIFs that only allows selecting GIFs from Tenor. The option for images will confusingly allow you to upload GIFs, but you don't find out until you've submitted your post that it will only show the first frame.

**Reply by shmerl** [reply]
If it loops the video anyway, then I don't see a point in using gifs there at all. Even others like Twitter convert gifs to video anyway if you upload it. Smooth looping sounds like player caching issue which can be fixed in whatever player they are using.

**Reply by jlandahl** [reply]
>I don't see a point in using gifs there at all

I agree, it's just that there are millions of GIFs out there, and people are used to uploading a GIF and seeing it animated automatically. They don't need to know or care if the platform converts the GIF to another format behind the scenes.

Bluesky really should do the conversion automatically like every other platform, especially when it's so easy to automate with FFMPEG or similar. (And especially since they accept GIFs but then surprisingly show only the first frame with no animation.)

**Reply by shmerl** [reply]
Yeah, I agree that conversion should be possible to implement. Does it upload video as is or it also transcodes it?

**Reply by jlandahl** [reply]
> Does it upload video as is or it also transcodes it?

There's a spinner with the text "Processing video..." when you upload an MP4, so I assume they have their own preferred optimizations that they apply to video uploads.

**Reply by shmerl** [reply]
Then I'm sure they can handle gifs the same way.

**Reply by tinspin** [reply]
Gifs are lossless = not blurry, ever. Also the dithering is an artistic choice. People that don't understand why gif > mp4 are usually not posting clips of pixel art games.

**Comment by batpad** [comment]
I made a simple tool that converts GIF to WebM in the browser (i.e. no upload required): https://batpad.github.io/gif-to-video/ which maybe helpful for folks who don't want to install video transcoding software or ffmpeg. The tool runs ffmpeg in the browser using [ffmpeg.wasm](https://github.com/ffmpegwasm/ffmpeg.wasm).

In general, I can see this being a hard decision on whether to support or not. GIFs are a pretty terrible format for moving images in general and the only reason they are in such widespread use is that browsers support them with the `<img>` tag, and for a lot of places that only allowed `<img>` tags, this was the only way to do animations.

Now that we have broad support for `<video>`, in general one can achieve much better compression and quality with video formats than GIF, and it is possible to have better programmatic control over `<video>` elements, i.e. controlling start / stop / pause / loop, etc. whereas GIFs will just always loop by default which might not always be what we want.

I think my ideal solution would be:

 - When the user is posting, allow them to select a GIF, but automatically convert it to a video format on the client, like in the example I linked above. So users can still "upload GIFs", but the server does not need to deal with the storage of a wasteful format and can use more efficient video formats, but this happens transparently to the user. The advantage of doing this client-side would be that one does not need to run complex backend video encoding infrastructure. I imagine this is a **lot** of complexity though and non-trivial to implement.

**Reply by vrad-exe** [reply]
Also, newer image formats which can do animation exist that are now supported by all major browsers. PNG and WEBP both support animation (though the former I believe is an unofficial extension of the specification, but browsers support it regardless). It would probably be better to focus on supporting one of those instead of GIF.

That being said, there are a lot of existing GIF files on the internet which people will try and upload, so it should definitely support auto-conversion on upload from GIF to whatever format is chosen.

**Reply by TodePond** [reply]
choice is good

**Reply by jlandahl** [reply]
> I made a simple tool that converts GIF to WebM in the browser (i.e. no upload required): https://batpad.github.io/gif-to-video/ which maybe helpful for folks who don't want to install video transcoding software or ffmpeg. The tool runs ffmpeg in the browser using [ffmpeg.wasm](https://github.com/ffmpegwasm/ffmpeg.wasm).

I wasn't aware of ffmpeg.wasm. That's really cool!

**Reply by tinspin** [reply]
COOP/COEP Service Worker not registered, a secure context is required.
FFmpeg load error: ReferenceError: SharedArrayBuffer is not defined.

So you need HTTPS?

Interesting difference to the threads generated mp4 (almost pixel perfect copy of the original gif):

https://multiplayeronlinestandard.com/monsters2.gif
https://multiplayeronlinestandard.com/monsters2.mp4
https://multiplayeronlinestandard.com/monsters2.webm

But what happens when you upload it to social platforms? My guess is re-encoded to oblivion? Most don't even allow .webm as upload file format! Maybe add a webm/mp4 drop down? Also can ffmpeg work on the main thread?

**Comment by KennedyRichard** [comment]
Didn't test it myself yet, but saw several posts claiming and confirming gifs can now be directly uploaded and are automatically converted into video.

Was that the final solution adopted or there's more to it (perhaps in the future)?

Just wanted to know cause didn't see anything like that mentioned here or in the Bluesky profile on the site.

Thanks in advance

**Reply by Zero3K** [reply]
I think it is what they are having it do since that is what Twitter does.

**Reply by mozzius** [reply]
The web client now accepts .gif files as source files for videos, yep!

**Reply by tinspin** [reply]
Still does not solve crisp pixel perfect animated .gifs for pixel art games which is mainly why this thread is here.

Blurry .mp4 is something we have been able to do forever...

**Reply by Zero3K** [reply]
Twitter doesn't upload GIFs as without converting them. They get converted to MP4 videos (such as https://video.twimg.com/tweet_video/GdkB5OFXQAAQW15.mp4) and don't display the progress bar and show a box with the word GIF in the lower left-hand corner with a play/pause button next to it.
